findoutlie/outfind.py

```diff
@@ -3,13 +3,108 @@
 
 import os.path as op
 from glob import glob
+from functools import partial
 
+import numpy as np
+import nibabel as nib
 
-def detect_outliers(fname):
-    return [42]
+from findoutlie.detectors import std_detector, iqr_detector
+from findoutlie.metrics import dvars
+from findoutlie.spm_funcs import get_spm_globals
+
+def detect_outliers(fname, metric='vol_means', detector='std', 
+                    n_stds=2, iqr_proportion=1.5, mask_join='or'):
+    """
+    Choose a timecourse metric and outlier detection method to apply to a given
+    file.
+
+    Parameters
+    ----------
+    fname : str, the timeseries file to be processed for outliers 
+
+    metric options : can be a str or list of options
+        The metric used for summarizing timecourse data across volumes before
+        applying outlier detecton method.
+
+        Options can include:
+        -'vol_means'
+        -'dvars'
+        -'spm_global'
+        -'all' (chooses all of the above)
+        -list of some or all of the above options
+
+    detector options : can be a str or list of options
+        The outlier detection method to use.
+
+        Options can include:
+        -'std'
+        -'iqr'
+        -'all' (chooses all of the above)
+        -list of some or all of the above options
+    
+    n_stds : int or float (default=2)
+        The standard deviation threshold used with the 'std' 
+        detector method.
+
+    iqr_proportion : int or float (default=1.5)
+        The interquartile proportion threshold to use with the
+        'iqr' method.
+
+    mask_join : str (default='or')
+        The method used to join outlier masks obtained when multiple
+        metrics or detection methods are used.
+
+        Options can be 'or' for np.logical_or or 'and' for np.logical_and.
+    """
+
+    metric_funcs = {
+        'vol_means': (lambda file: np.mean(nib.load(file).get_fdata(), axis=(0,1,2))),
+        'dvars': (lambda file: dvars(nib.load(file))),
+        'spm_global': (lambda file: np.array(get_spm_globals(file)))
+    }
+
+    detector_funcs = {
+        'std': partial(std_detector, n_stds=n_stds),
+        'iqr': partial(iqr_detector, iqr_proportion=iqr_proportion)
+    }
+    assert type(metric) in [list, str], "'metric' must be a str or list"
+    assert type(detector) in [list, str], "'detector' must be str or list"
+    assert mask_join in ['or', 'and'], "'mask_join' must be 'or' or 'and'"
+
+    if mask_join == 'or':
+        mask_join = np.logical_or
+    elif mask_join == 'and':
+        mask_join = np.logical_and
+    
+    if type(metric) == str:
+        if metric == 'all':
+            metric = list(metric_funcs.keys())
```

Comment:
This also works (`.keys()` is the default result of interating on a dictionary):

```suggestion
            metric = list(metric_funcs)
```